platform/default/mbgl/map/map_snapshotter.cpp

```diff
@@ -0,0 +1,76 @@
+#include <mbgl/map/map_snapshotter.hpp>
+
+#include <mbgl/actor/actor_ref.hpp>
+#include <mbgl/gl/headless_frontend.hpp>
+#include <mbgl/map/map.hpp>
+#include <mbgl/storage/file_source.hpp>
+#include <mbgl/style/style.hpp>
+#include <mbgl/util/event.hpp>
+
+namespace mbgl {
+
+class MapSnapshotter::Impl {
+public:
+    Impl(FileSource&,
+         Scheduler&,
+         const std::string& styleURL,
+         const Size&,
+         const float pixelRatio,
+         const CameraOptions&,
+         const optional<LatLngBounds> region,
+         const optional<std::string> programCacheDir);
+
+    void snapshot(ActorRef<MapSnapshotter::Callback>);
+
+private:
+    HeadlessFrontend frontend;
+    Map map;
+};
+
+MapSnapshotter::Impl::Impl(FileSource& fileSource,
+           Scheduler& scheduler,
+           const std::string& styleURL,
+           const Size& size,
+           const float pixelRatio,
+           const CameraOptions& cameraOptions,
+           const optional<LatLngBounds> region,
+           const optional<std::string> programCacheDir)
+    : frontend(size, pixelRatio, fileSource, scheduler, programCacheDir)
+    , map(frontend, MapObserver::nullObserver(), size, pixelRatio, fileSource, scheduler, MapMode::Still) {
+
+    map.getStyle().loadURL(styleURL);
+
+    map.jumpTo(cameraOptions);
+
+    // Set region, if specified
+    if (region) {
+        mbgl::EdgeInsets insets = { 0, 0, 0, 0 };
+        std::vector<LatLng> latLngs = { region->southwest(), region->northeast() };
+        map.jumpTo(map.cameraForLatLngs(latLngs, insets));
+    }
+}
+
+void MapSnapshotter::Impl::snapshot(ActorRef<MapSnapshotter::Callback> callback) {
+    map.renderStill([this, callback = std::move(callback)] (std::exception_ptr error) mutable {
+        callback.invoke(&MapSnapshotter::Callback::operator(), error, error ? PremultipliedImage() : frontend.readStillImage());
+    });
+}
+
+MapSnapshotter::MapSnapshotter(FileSource& fileSource,
+                               Scheduler& scheduler,
+                               const std::string& styleURL,
+                               const Size& size,
+                               const float pixelRatio,
+                               const CameraOptions& cameraOptions,
+                               const optional<LatLngBounds> region,
+                               const optional<std::string> programCacheDir)
+   : impl(std::make_unique<util::Thread<MapSnapshotter::Impl>>("Map Snapshotter", fileSource, scheduler, styleURL, size, pixelRatio, cameraOptions, region, programCacheDir)) {
+}
+
+MapSnapshotter::~MapSnapshotter() = default;
```

Comment:
This destructor synchronously destroys the thread. If a render is in progress, this could potentially take a few dozen or even hundred milliseconds. We could introduce a similar system like the tile parsing `obsolete` atomic bool that we check after every layer and abort rendering in case we try to abort the rendering.

Reply:
Could be worth looking into. This doesn't add too much overhead in the "regular" case, it does add a check every frame for every layer in Still and Continuous mode.